clang/test/CodeGenSYCL/check-direct-attribute-propagation.cpp

```diff
@@ -0,0 +1,319 @@
+
+template <int SIZE>
+class Functor1 {
+public:
+  [[intel::num_simd_work_items(SIZE)]] void operator()() const {}
+};
+
+[[intel::num_simd_work_items(5)]] void foo1() {}
+
+class Foo2 {
+public:
+  [[intel::no_global_work_offset(1)]] void operator()() const {}
+};
+
+template <int SIZE>
+class Functor2 {
+public:
+  [[intel::no_global_work_offset(SIZE)]] void operator()() const {}
+};
+
+[[intel::no_global_work_offset(0)]] void foo2() {}
+
+class Foo3 {
+public:
+  [[intel::max_global_work_dim(1)]] void operator()() const {}
+};
+
+template <int SIZE>
+class Functor3 {
+public:
+  [[intel::max_global_work_dim(SIZE)]] void operator()() const {}
+};
+
+[[intel::max_global_work_dim(1)]] void foo3() {}
+
+
+class Foo4 {
+public:
+  [[sycl::reqd_sub_group_size(16)]] void operator()() const {}
+};
+
+[[sycl::reqd_sub_group_size(8)]] void foo4() {}
+
+class Functor4 {
+public:
+  void operator()() const {
+    foo();
+  }
+};
+
+template <int SIZE>
+class Functor5 {
+public:
+  [[sycl::reqd_sub_group_size(SIZE)]] void operator()() const {}
+};
+
+class Foo5 {
+public:
+  [[sycl::reqd_work_group_size(32, 16, 16)]] void operator()() const {}
+};
+
+[[sycl::reqd_work_group_size(8, 1, 1)]] void foo5() {}
+
+class Functor6 {
+public:
+  void operator()() const {
+    foo5();
+  }
+};
+
+template <int SIZE, int SIZE1, int SIZE2>
+class Functor7 {
+public:
+  [[sycl::reqd_work_group_size(SIZE, SIZE1, SIZE2)]] void operator()() const {}
+};
+
+class Foo6 {
+public:
+  [[intel::max_work_group_size(32, 16, 16)]] void operator()() const {}
+};
+
+[[intel::max_work_group_size(8, 1, 1)]] void foo6() {}
+
+class Functor8 {
+public:
+  void operator()() const {
+    foo6();
+  }
+};
+
+template <int SIZE, int SIZE1, int SIZE2>
+class Functor9 {
+public:
+  [[intel::max_work_group_size(SIZE, SIZE1, SIZE2)]] void operator()() const {}
+};
+
+class Foo7 {
+public:
+  [[intel::sycl_explicit_simd]] void operator()() const {}
+};
+
+[[intel::sycl_explicit_simd]] void foo7() {}
+
+class Foo8 {
+public:
+  [[intel::kernel_args_restrict]] void operator()() const {}
+};
+
+[[intel::kernel_args_restrict]] void foo8() {}
+
+class Functor10 {
+public:
+  void operator()() const {
+    foo8();
+  }
+};
+
+int main() {
+  q.submit([&](handler &h) {
+    // CHECK: define {{.*}}spir_kernel void @{{.*}}kernel_name1() #0 {{.*}} !scheduler_target_fmax_mhz ![[NUM1:[0-9]+]]
+    Foo boo;
+    h.single_task<class kernel_name1>(boo);
+
+    // CHECK: define {{.*}}spir_kernel void @{{.*}}kernel_name2() #0 {{.*}} !scheduler_target_fmax_mhz ![[NUM42:[0-9]+]]
+    h.single_task<class kernel_name2>(
+        []() [[intel::scheduler_target_fmax_mhz(42)]]{});
+
+    // CHECK: define {{.*}}spir_kernel void @{{.*}}kernel_name3() #0 {{.*}} !scheduler_target_fmax_mhz ![[NUM2:[0-9]+]]
+    Functor<2> f;
+    h.single_task<class kernel_name3>(f);
+
+    // Test attribute is not propagated.
+    // CHECK: define {{.*}}spir_kernel void @{{.*}}kernel_name4()
+    // CHECK-NOT: !scheduler_target_fmax_mhz
```

Comment:
>>This test will pass even if scheduler_target_fmax_mhz is present in define {{.*}}spir_kernel void @{{.*}}kernel_name4() right? 

No, it will not pass.

I have updated tests with CHECK-SAME.